### DISCUSSION TITLE
datasets/classify/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#10067] [Docs]
# datasets/classify/

Explore image classification datasets supported by Ultralytics, learn the standard dataset format, and set up your own dataset for training models.

https://docs.ultralytics.com/datasets/classify/

**Comment by wynshiter** [comment]
How to obtain a portion of data from a big dataset during classification training?
For example, my dataset itself comes from many places, and I want to test the final impact of different proportions of data labels on the results. So the movement of the data itself is quite cumbersome. Would it be better if the trained YAML could support data paths? How can we achieve this

**Reply by glenn-jocher** [reply]
Hey there! 👋 It's great to hear you're exploring different portions of your dataset for classification training. To test the impact of various data label proportions without physically moving your data, you can indeed use a customized YAML file to specify different data paths. This can simplify managing and experimenting with subsets of your dataset. 

For your specific scenario, you might consider creating a smaller dataset directory following the structure outlined in our [docs](https://docs.ultralytics.com/datasets/classify/), where each class subdirectory contains only a portion of your data. Then, point your training YAML to this new, smaller dataset. Here's a quick example of how your YAML file could look:

```yaml
path: path/to/your/smaller_dataset
train: train/
val: val/
test: test/  # optional
nc: 10  # number of classes
names: ['class_1', 'class_2', ..., 'class_n']
```

And when you train, just reference this YAML file:

```bash
yolo detect train data=smaller_dataset.yaml model=yolov8n-cls.pt epochs=100 imgsz=640
```

This approach allows you to easily switch between different dataset proportions by creating various YAML files for each scenario without moving your actual data. Happy experimenting! 🚀

**Reply by rromanchuk** [reply]
lol @ the man hours wasted on this repo from LLMs just making shit up

**Comment by clementguerin1** [comment]
Hi,

In which folder should my label go ?  

something like this ?

cifar-10-/
|
|-- train/
|   |-- airplane/
|   |   |-- 10008_airplane.png
|   |   |-- 10008_airplane.txt
|   |   |-- 10009_airplane.png
|   |   |-- 10009_airplane.txt

my txt file look like this 
6 0.73359375 0.4921875 0.0703125 0.18515625
class + bouding box

**Reply by pderrenger** [reply]
@clementguerin1 hi there,

Thanks for reaching out! For classification tasks with YOLOv8, your dataset structure looks almost correct. However, you should not include the `.txt` label files in your dataset as YOLOv8 classification tasks don't require separate label files. Each subdirectory name acts as the label for the images contained within.

Here's how you should organize your data:

```
cifar-10-/
|
|-- train/
|   |-- airplane/
|   |   |-- 10008_airplane.png
|   |   |-- 10009_airplane.png
|   |-- automobile/
|   |   |-- 10010_automobile.png
|   |   |-- 10011_automobile.png
|
|-- test/
|   |-- airplane/
|   |   |-- 20008_airplane.png
|   |   |-- 20009_airplane.png
|   |-- automobile/
|   |   |-- 20010_automobile.png
|   |   |-- 20011_automobile.png
|
|-- val/ (optional)
|   |-- airplane/
|   |   |-- 30008_airplane.png
|   |   |-- 30009_airplane.png
|   |-- automobile/
|   |   |-- 30010_automobile.png
|   |   |-- 30011_automobile.png
```

Make sure each image is placed in the folder named after its class. This way, the model will automatically derive classes from directory names, making .txt files unnecessary. Best of luck with your training! 🚀

**Comment by Jack-Wong94** [comment]
How do I include background image to the dataset (i.e. image that does not belong to any of the classes)?

**Reply by glenn-jocher** [reply]
To include background images that do not belong to any class in your dataset, you can create a separate class directory named "background" and place all such images there. This way, the model will learn to recognize these images as a distinct class.

Here's how you can structure your dataset:

```
dataset/
|
|-- train/
|   |-- airplane/
|   |-- automobile/
|   |-- bird/
|   |-- background/  # Add this directory for background images
|
|-- test/
|   |-- airplane/
|   |-- automobile/
|   |-- bird/
|   |-- background/  # Add this directory for background images
|
|-- val/ (optional)
|   |-- airplane/
|   |-- automobile/
|   |-- bird/
|   |-- background/  # Add this directory for background images
```

By following this structure, you can ensure that the model is trained to distinguish between your target classes and background images. For more details on dataset structure, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/datasets/classify/). Happy training! 🚀

**Reply by MagicLogicYourDad** [reply]
Thanks YOLO, it is incredibly powerful!!!

However, I still have some questions. As you suggested, I added a backgrounddirectory under both the trainand valdirectories, and stored background images inside. I'm unsure if this is working correctly, but when I look at the charts in the training folder, for example /trainX/confusion_matrix_normalized.png, I see two backgroundentries on the x and y axes. One backgroundhas data, while the other does not. Does this mean YOLO sees a class named background, rather than truly recognizing it as background (negative samples)? I look forward to your reply.

**Comment by mopa91** [comment]
Hello,
I have trained a classification model with my custom dataset. Now, I want to validate the model on another dataset. How can I use this: validation_results = model.val(), to validate the model's predictions and get the confusion matrix for the predicted data?

**Reply by pderrenger** [reply]
Hello,

Great to hear that you've trained a classification model with your custom dataset! To validate your model on another dataset and obtain a confusion matrix, you can use the `model.val()` method. Here's a concise example:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO("path/to/your/model.pt")

# Validate the model on a new dataset
validation_results = model.val(data="path/to/validation/dataset")

# Access the confusion matrix
confusion_matrix = validation_results.confusion_matrix
print(confusion_matrix)
```

Ensure your validation dataset follows the required directory structure. For more details on dataset organization and validation, you can refer to the [Image Classification Datasets Overview](https://docs.ultralytics.com/datasets/classify/).

Happy validating! 🚀

**Reply by momint2020** [reply]
Hello,

Thank you very much for your reply. It was a great help. Could you please tell me how I can get the F1 score curve for my classification model on the custom dataset?

**Reply by pderrenger** [reply]
@momint2020 hello,

I'm glad to hear the previous information was helpful! To generate an F1 score curve for your classification model on a custom dataset, you can use the `val()` method to perform validation and then analyze the predictions to calculate F1 scores at various thresholds.

Here’s a basic outline of how you might approach this:

1. **Validate the Model**: First, run validation to get the predictions.
   ```python
   from ultralytics import YOLO

   # Load your trained model
   model = YOLO("path/to/your/model.pt")

   # Validate the model on your custom dataset
   validation_results = model.val(data="path/to/your/validation/dataset")
   ```

2. **Calculate F1 Scores**: You'll need to write a custom function or script to calculate the F1 score for each class at different thresholds, using the predictions obtained. This involves calculating precision and recall for each threshold and then using these to calculate the F1 score.

3. **Plot the Curve**: Use a plotting library like `matplotlib` to plot the F1 scores against the thresholds.

Unfortunately, direct support for plotting F1 curves isn't built into the Ultralytics framework, so this will require some custom implementation on your part.

If you need further guidance on the specifics of calculating and plotting the F1 score, please let me know!

**Comment by casperking666** [comment]
for classification task, do I still need the data.yaml file? If not, how should I carry out validation and testing using the folder structure shown above?

**Reply by glenn-jocher** [reply]
@casperking666 hi there!

For classification tasks with Ultralytics YOLO, you do not need a `data.yaml` file. Instead, you should organize your dataset in the specific folder structure as outlined in the documentation. This structure includes separate directories for training (`train`), testing (`test`), and optionally validation (`val`), with each class having its own subdirectory containing the relevant images.

Here’s a quick example of how your dataset should be structured:

```
dataset/
|
|-- train/
|   |-- class1/
|   |   |-- image1.jpg
|   |   |-- image2.jpg
|   |   |-- ...
|   |
|   |-- class2/
|   |   |-- image1.jpg
|   |   |-- image2.jpg
|   |   |-- ...
|   |
|   |-- ...
|
|-- test/
|   |-- class1/
|   |   |-- image1.jpg
|   |   |-- image2.jpg
|   |   |-- ...
|   |
|   |-- class2/
|   |   |-- image1.jpg
|   |   |-- image2.jpg
|   |   |-- ...
|   |
|   |-- ...
|
|-- val/ (optional)
|   |-- class1/
|   |   |-- image1.jpg
|   |   |-- image2.jpg
|   |   |-- ...
|   |
|   |-- class2/
|   |   |-- image1.jpg
|   |   |-- image2.jpg
|   |   |-- ...
|   |
|   |-- ...
```

To carry out validation and testing, simply ensure that your dataset is organized as shown above. The `train`, `test`, and `val` directories should each contain subdirectories named after the classes, with the images for each class inside these subdirectories.

Here’s how you can train and validate your model using the Ultralytics YOLO API:

### Python Example
```python
from ultralytics import YOLO

# Load a pretrained classification model
model = YOLO("yolov8n-cls.pt")

# Train the model
results = model.train(data="path/to/dataset", epochs=100, imgsz=640)

# Validate the model
metrics = model.val(data="path/to/dataset")
```

### CLI Example
```bash
# Train the model
yolo classify train data=path/to/dataset model=yolov8n-cls.pt epochs=100 imgsz=640

# Validate the model
yolo classify val data=path/to/dataset model=yolov8n-cls.pt
```

For more detailed information, you can refer to the [Ultralytics documentation on classification datasets](https://docs.ultralytics.com/datasets/classify/).

If you encounter any issues, please ensure you are using the latest version of the packages. If the problem persists, providing a minimum reproducible example can greatly help in diagnosing the issue. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Happy training! 🚀

**Reply by casperking666** [reply]
thanks! I used comet ml during training, and it produced plots metrics/accuracy_top1 VS step and metrics/accuracy_top5 VS step. What sets did they use to generate these plots? I couldn't find any label indicate what set that is. I would assume it's the training set, as the accuracy is getting to close to 100%.

second question, it's not clear to me what set this line of code is using metrics = model.val(data="path/to/dataset")? Is it the validation set or the testing set? If it's the validation set, how would I test the test set as they are all organized under the same directory as you previously showed?

**Reply by glenn-jocher** [reply]
Hi @casperking666,

Thank you for your questions! Let's address them one by one:

1. **Comet ML Metrics:**
   The plots for `metrics/accuracy_top1 VS step` and `metrics/accuracy_top5 VS step` generated by Comet ML typically use the validation set during training. This is because the validation set provides a more accurate measure of the model's performance on unseen data, which helps in monitoring overfitting. If the accuracy is approaching 100%, it might indicate that the model is overfitting to the training data. To confirm, you can check the Comet ML experiment logs or the training script to see which dataset was used for these metrics.

2. **Validation and Testing Sets:**
   The `metrics = model.val(data="path/to/dataset")` line of code uses the validation set if it exists. If no validation set is provided, it defaults to using the test set. To explicitly validate or test your model, ensure your dataset is organized as follows:
   
   ```
   dataset/
   ├── train/
   │   ├── class1/
   │   ├── class2/
   │   └── ...
   ├── test/
   │   ├── class1/
   │   ├── class2/
   │   └── ...
   └── val/ (optional)
       ├── class1/
       ├── class2/
       └── ...
   ```

   If you want to validate using the test set, you can specify the `split` argument in the `val` method:
   
   ```python
   metrics = model.val(data="path/to/dataset", split='test')
   ```

   This ensures that the model uses the test set for validation.

If you encounter any issues, please ensure you are using the latest version of the Ultralytics packages. Providing a minimum reproducible example can greatly help in diagnosing the issue. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions. Happy training! 🚀

**Reply by casperking666** [reply]
I see, thank you! I have tried to train from scratch using the yaml file like the following:
```python
model = YOLO("yolov8n-cls.yaml")
```
but I got the following error
```code
Traceback (most recent call last):
  File "/user/work/yf20630/cow-dataset-project/train.py", line 27, in <module>
    model.save('./models/yolov8s-cow-crops-2080-1e.pt')
  File "/user/work/yf20630/miniconda3/envs/cow_project/lib/python3.10/site-packages/ultralytics/engine/model.py", line 335, in save
    torch.save({**self.ckpt, **updates}, filename, use_dill=use_dill)
TypeError: 'NoneType' object is not a mapping
```
on this line
```python
model.save('./models/yolov8s-cow-crops-2080-1e.pt')
```

the saving works perfectly fine when using the pretrained model, but wouldn't work here, any thoughts on why?

**Reply by glenn-jocher** [reply]
Hi @casperking666,

Thank you for reaching out and providing the details of your issue!

It looks like you're encountering an error when trying to save a model that was initialized from a YAML configuration file. The error message `TypeError: 'NoneType' object is not a mapping` suggests that there might be an issue with the model's checkpoint data.

Here are a few steps to help resolve this:

1. **Ensure Proper Initialization:**
   When initializing a model from a YAML file, make sure that the YAML file is correctly configured and that the model is properly built before attempting to save it. Here’s an example of how to initialize and train the model:

   ```python
   from ultralytics import YOLO

   # Initialize a model from a YAML file
   model = YOLO("yolov8n-cls.yaml")

   # Train the model
   results = model.train(data="path/to/dataset", epochs=100, imgsz=640)

   # Save the trained model
   model.save('./models/yolov8s-cow-crops-2080-1e.pt')
   ```

2. **Verify Training Completion:**
   Ensure that the training process completes successfully before saving the model. If the training is interrupted or fails, the model's checkpoint data might not be properly populated.

3. **Check for Updates:**
   Verify that you are using the latest version of the Ultralytics package. Sometimes, issues are resolved in newer releases. You can update the package using:
   ```bash
   pip install --upgrade ultralytics
   ```

4. **Provide a Reproducible Example:**
   If the issue persists, it would be helpful to provide a minimum reproducible example. This will allow us to better understand the context and provide more accurate assistance. You can find guidelines for creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

By following these steps, you should be able to identify and resolve the issue. If you continue to experience problems, please share more details or a reproducible example, and we’ll be happy to assist further.

Best of luck with your training! 🚀

Feel free to reach out if you have any more questions.

**Reply by MrLi-creator** [reply]
This is my own small_data.yaml file:
path: C:/Users/AIX1WX/Documents/10-POC/train_yolov8/yolov8-all-in-one-main_classification/datasets
train: train/
val: val/
test: test/

nc: 7
names: ['ABS','Airbag','Background','EPB','EPBGREEN','EPBRed','QMtt']

class_weights: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 3.0]

And my train coda here:
from ultralytics import YOLO
import os
os.environ["KMP_DUPLICATE_LIB_OK"]="TRUE"
if __name__ == '__main__': 
    model = YOLO("yolov8n-cls.pt")  # load a pretrained model (recommended for training)
    results = model.train(data="small_data.yaml", epochs=30, imgsz=50, device=None, batch=16)

Why will report the error:
Dataset not found , missing path C:\Users\AIX1WX\Documents\10-POC\train_yolov8\yolov8-all-in-one-main_detection\datasets\small_data.yaml, attempting download...
Downloading https://github.com/ultralytics/yolov5/releases/download/v1.0/small_data.yaml.zip to 'C:\Users\AIX1WX\Documents\10-POC\train_yolov8\yolov8-all-in-one-main_detection\datasets\small_data.yaml.zip'...

**Reply by glenn-jocher** [reply]
@MrLi-creator it looks like the error is due to the dataset path not being correctly recognized. Ensure that the path in your `small_data.yaml` file is correct and accessible. The path should be absolute or relative to your current working directory. Also, make sure the dataset structure follows the required format with `train`, `val`, and `test` directories. If the issue persists, please verify it with the latest version of the Ultralytics package. If you need further assistance, feel free to ask!

**Comment by MrLi-creator** [comment]
This is my own small_data.yaml file:
path: C:/Users/AIX1WX/Documents/10-POC/train_yolov8/yolov8-all-in-one-main_classification/datasets
train: train/
val: val/
test: test/

nc: 7
names: ['ABS','Airbag','Background','EPB','EPBGREEN','EPBRed','QMtt']

class_weights: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 3.0]

And my train coda here:
    from ultralytics import YOLO
    import os
    os.environ["KMP_DUPLICATE_LIB_OK"]="TRUE"
    if __name__ == '__main__':
        model = YOLO("yolov8n-cls.pt") # load a pretrained model (recommended for training)
        results = model.train(data="small_data.yaml", epochs=30, imgsz=50, device=None, batch=16)

Why will report the error:
Dataset not found , missing path C:\Users\AIX1WX\Documents\10-POC\train_yolov8\yolov8-all-in-one-main_detection\datasets\small_data.yaml, attempting download...
Downloading https://github.com/ultralytics/yolov5/releases/download/v1.0/small_data.yaml.zip to 'C:\Users\AIX1WX\Documents\10-POC\train_yolov8\yolov8-all-in-one-main_detection\datasets\small_data.yaml.zip'...

**Reply by glenn-jocher** [reply]
@MrLi-creator it looks like the error is due to the dataset path specified in your `small_data.yaml` file. The path should be relative to the location of the `small_data.yaml` file or an absolute path. Ensure that the dataset directory structure follows the required format, with `train`, `val`, and `test` subdirectories, each containing subdirectories for each class. Verify that the path in your `small_data.yaml` is correct and points to the dataset's root directory. For more details on dataset structure, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/datasets/classify/).

**Reply by MrLi-creator** [reply]
Hi glenn-jocher,
I find the case: I first train __yolov8n.pt__ in yolov8-all-in-one-main_detection file. And second time I train __yolov8n-cls.pt__ in another file yolov8-all-in-one-main_classification. Then everytime train __yolov8n-cls.pt__, the dataset path always points to yolov8-all-in-one-main_detection. So is there any way to change the default path of the dataset

**Reply by glenn-jocher** [reply]
@MrLi-creator it seems like the issue arises from the dataset path being cached or incorrectly set. To resolve this, ensure that the `data` argument in your training script points to the correct `small_data.yaml` file. Additionally, verify that the paths specified within `small_data.yaml` are accurate and accessible. If the problem persists, try clearing any cached paths or restarting your training environment. For further details on dataset structure, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/datasets/classify/).

**Reply by MrLi-creator** [reply]
Hi glenn-jocher,
I just find that the classification task can't use data.yaml. You can see pderrenger comment in the link(https://github.com/ultralytics/ultralytics/issues/13789)

**Reply by glenn-jocher** [reply]
Thank you for pointing that out. For classification tasks with YOLO, the dataset should be organized in a specific directory structure rather than using a `data.yaml` file. Ensure your dataset follows the format with separate `train`, `val`, and `test` directories, each containing subdirectories for each class. This structure is essential for proper training and evaluation. For more details, you can refer to the [Ultralytics documentation on dataset structure](https://docs.ultralytics.com/datasets/classify/).

**Comment by leolle520** [comment]
Can I use this model to train an one class image classifier as an abnormal image detector? Thank you.

**Reply by glenn-jocher** [reply]
Yes, you can use the YOLO model to train a one-class image classifier for abnormal image detection. Simply organize your dataset with one class for normal images and another for abnormal images, then proceed with training. For more details on dataset structure, visit our documentation.

**Comment by NajeebAhmad00** [comment]
I am using a yolov8-cls.pt model and want to add a new class to it. I need to train the model on this new class without affecting the existing classes. How can I do this?

**Reply by glenn-jocher** [reply]
@NajeebAhmad00 to add a new class to your `yolov8-cls.pt` model without affecting existing classes, you can use transfer learning. Organize your dataset following the structure outlined in the [Ultralytics documentation](https://docs.ultralytics.com/datasets/classify/), then fine-tune the model on your new dataset. Here's an example in Python:

```python
from ultralytics import YOLO

# Load the pretrained model
model = YOLO("yolov8n-cls.pt")

# Train the model on the new dataset
results = model.train(data="path/to/your/dataset", epochs=100, imgsz=640)
```

This will allow the model to learn the new class while retaining knowledge of the existing ones.

**Comment by rishabh-jain14** [comment]
I have this yaml file for caltech-256 classification:
path: D:/Coursework/Fall 24/CSC591-007 DLBA/Project/Project/data/tvt_Caltech_256
train: train/
val: val/
test: test/
nc: 257
names: ['001.ak47', '002.american-flag', '003.backpack', '004.baseball-bat', '005.baseball-glove', '006.basketball-hoop', '007.bat', '008.bathtub', '009.bear', '010.beer-mug', '011.billiards', '012.binoculars', '013.birdbath', '014.blimp', '015.bonsai-101', '016.boom-box', '017.bowling-ball', '018.bowling-pin', '019.boxing-glove', '020.brain-101', '021.breadmaker', '022.buddha-101', '023.bulldozer', '024.butterfly', '025.cactus', '026.cake', '027.calculator', '028.camel', '029.cannon', '030.canoe', '031.car-tire', '032.cartman', '033.cd', '034.centipede', '035.cereal-box', '036.chandelier-101', '037.chess-board', '038.chimp', '039.chopsticks', '040.cockroach', '041.coffee-mug', '042.coffin', '043.coin', '044.comet', '045.computer-keyboard', '046.computer-monitor', '047.computer-mouse', '048.conch', '049.cormorant', '050.covered-wagon', '051.cowboy-hat', '052.crab-101', '053.desk-globe', '054.diamond-ring', '055.dice', '056.dog', '057.dolphin-101', '058.doorknob', '059.drinking-straw', '060.duck', '061.dumb-bell', '062.eiffel-tower', '063.electric-guitar-101', '064.elephant-101', '065.elk', '066.ewer-101', '067.eyeglasses', '068.fern', '069.fighter-jet', '070.fire-extinguisher', '071.fire-hydrant', '072.fire-truck', '073.fireworks', '074.flashlight', '075.floppy-disk', '076.football-helmet', '077.french-horn', '078.fried-egg', '079.frisbee', '080.frog', '081.frying-pan', '082.galaxy', '083.gas-pump', '084.giraffe', '085.goat', '086.golden-gate-bridge', '087.goldfish', '088.golf-ball', '089.goose', '090.gorilla', '091.grand-piano-101', '092.grapes', '093.grasshopper', '094.guitar-pick', '095.hamburger', '096.hammock', '097.harmonica', '098.harp', '099.harpsichord', '100.hawksbill-101', '101.head-phones', '102.helicopter-101', '103.hibiscus', '104.homer-simpson', '105.horse', '106.horseshoe-crab', '107.hot-air-balloon', '108.hot-dog', '109.hot-tub', '110.hourglass', '111.house-fly', '112.human-skeleton', '113.hummingbird', '114.ibis-101', '115.ice-cream-cone', '116.iguana', '117.ipod', '118.iris', '119.jesus-christ', '120.joy-stick', '121.kangaroo-101', '122.kayak', '123.ketch-101', '124.killer-whale', '125.knife', '126.ladder', '127.laptop-101', '128.lathe', '129.leopards-101', '130.license-plate', '131.lightbulb', '132.light-house', '133.lightning', '134.llama-101', '135.mailbox', '136.mandolin', '137.mars', '138.mattress', '139.megaphone', '140.menorah-101', '141.microscope', '142.microwave', '143.minaret', '144.minotaur', '145.motorbikes-101', '146.mountain-bike', '147.mushroom', '148.mussels', '149.necktie', '150.octopus', '151.ostrich', '152.owl', '153.palm-pilot', '154.palm-tree', '155.paperclip', '156.paper-shredder', '157.pci-card', '158.penguin', '159.people', '160.pez-dispenser', '161.photocopier', '162.picnic-table', '163.playing-card', '164.porcupine', '165.pram', '166.praying-mantis', '167.pyramid', '168.raccoon', '169.radio-telescope', '170.rainbow', '171.refrigerator', '172.revolver-101', '173.rifle', '174.rotary-phone', '175.roulette-wheel', '176.saddle', '177.saturn', '178.school-bus', '179.scorpion-101', '180.screwdriver', '181.segway', '182.self-propelled-lawn-mower', '183.sextant', '184.sheet-music', '185.skateboard', '186.skunk', '187.skyscraper', '188.smokestack', '189.snail', '190.snake', '191.sneaker', '192.snowmobile', '193.soccer-ball', '194.socks', '195.soda-can', '196.spaghetti', '197.speed-boat', '198.spider', '199.spoon', '200.stained-glass', '201.starfish-101', '202.steering-wheel', '203.stirrups', '204.sunflower-101', '205.superman', '206.sushi', '207.swan', '208.swiss-army-knife', '209.sword', '210.syringe', '211.tambourine', '212.teapot', '213.teddy-bear', '214.teepee', '215.telephone-box', '216.tennis-ball', '217.tennis-court', '218.tennis-racket', '219.theodolite', '220.toaster', '221.tomato', '222.tombstone', '223.top-hat', '224.touring-bike', '225.tower-pisa', '226.traffic-light', '227.treadmill', '228.triceratops', '229.tricycle', '230.trilobite-101', '231.tripod', '232.t-shirt', '233.tuning-fork', '234.tweezer', '235.umbrella-101', '236.unicorn', '237.vcr', '238.video-projector', '239.washing-machine', '240.watch-101', '241.waterfall', '242.watermelon', '243.welding-mask', '244.wheelbarrow', '245.windmill', '246.wine-bottle', '247.xylophone', '248.yarmulke', '249.yo-yo', '250.zebra', '251.airplanes-101', '252.car-side-101', '253.faces-easy-101', '254.greyhound', '255.tennis-shoes', '256.toad', '257.clutter']

However, when I train, it shows:
train: Scanning D:\Coursework\Fall 24\CSC591-007 DLBA\Project\Project\data\tvt_Caltech_256\train\001... 0 images, 24485 backgrounds, 0 corrupt: 100%|██████████| 24485/24485 [00:43<00:00, 565.92it/s]
train: WARNING  No labels found in D:\Coursework\Fall 24\CSC591-007 DLBA\Project\Project\data\tvt_Caltech_256\train\001.cache. See https://docs.ultralytics.com/datasets for dataset formatting guidance.

train: New cache created: D:\Coursework\Fall 24\CSC591-007 DLBA\Project\Project\data\tvt_Caltech_256\train\001.cache
WARNING  No labels found in D:\Coursework\Fall 24\CSC591-007 DLBA\Project\Project\data\tvt_Caltech_256\train\001.cache, training may not work correctly. See https://docs.ultralytics.com/datasets for dataset formatting guidance.
val: Scanning D:\Coursework\Fall 24\CSC591-007 DLBA\Project\Project\data\tvt_Caltech_256\val\001... 0 images, 3061 backgrounds, 0 corrupt: 100%|██████████| 3061/3061 [00:02<00:00, 1231.29it/s]
val: WARNING  No labels found in D:\Coursework\Fall 24\CSC591-007 DLBA\Project\Project\data\tvt_Caltech_256\val\001.cache. See https://docs.ultralytics.com/datasets for dataset formatting guidance.

val: New cache created: D:\Coursework\Fall 24\CSC591-007 DLBA\Project\Project\data\tvt_Caltech_256\val\001.cache
WARNING  No labels found in D:\Coursework\Fall 24\CSC591-007 DLBA\Project\Project\data\tvt_Caltech_256\val\001.cache, training may not work correctly. See https://docs.ultralytics.com/datasets for dataset formatting guidance.
Plotting labels to runs\detect\train\labels.jpg... 
zero-size array to reduction operation maximum which has no identity
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=3.8e-05, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
TensorBoard: model graph visualization added 
Image sizes 128 train, 128 val

What should I do to resolve the issue?

**Reply by glenn-jocher** [reply]
@rishabh-jain14 it looks like the issue stems from the dataset structure not aligning with the expected format for classification tasks. Each class should have its own directory containing the respective images. Please ensure your dataset follows this structure. For detailed guidance, refer to the [Image Classification Datasets Overview](https://docs.ultralytics.com/datasets/classify/).

**Comment by lauravchrobak** [comment]
Its confusing to me that the val set is optional for classification but the test set is optional for detection. Should there be some consistency there?

**Reply by glenn-jocher** [reply]
@lauravchrobak great observation! The difference in required splits stems from task-specific conventions. For classification, the test set is mandatory to ensure standardized evaluation of final accuracy, while val remains optional for hyperparameter tuning. Detection tasks often prioritize validation during training (like early stopping), hence the flexibility. This structure aligns with common practices in each domain while maintaining framework adaptability.

**Comment by iongion** [comment]
My annotations have regions that I know variate a lot and cannot be used for classification. Is there a way to feed these regions so that it avoids learning from them during training ? If I mask/fill the areas with black color on the image for training itself, wouldn't that affect classification ? Sometimes some images might have black shapes as features themselves.
